File: mopub_flutter-1.2.2+1/android/src/main/java/com/mopub_flutter/MopubInterstitialAdPlugin.java
package com.mopub_flutter;

import androidx.annotation.NonNull;

import com.mopub.mobileads.MoPubErrorCode;
import com.mopub.mobileads.MoPubInterstitial;

import java.util.HashMap;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;

class MopubInterstitialAdPlugin implements MethodChannel.MethodCallHandler {

    private HashMap<String, MoPubInterstitial> ads = new HashMap<>();

    private PluginRegistry.Registrar registrar;

    MopubInterstitialAdPlugin(PluginRegistry.Registrar registrar) {
        this.registrar = registrar;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        final HashMap args = (HashMap) call.arguments;
        final String adUnitId = (String) args.get("adUnitId");
        MoPubInterstitial ad;

        switch (call.method) {
            case MopubConstants.SHOW_INTERSTITIAL_METHOD:
                ad = ads.get(adUnitId);
                if (ad == null || !ad.isReady()) {
                    result.success(null);
                    return;
                }
                ad.show();
                result.success(null);
                break;
            case MopubConstants.LOAD_INTERSTITIAL_METHOD:
                if (registrar.activity() == null)
                    return;
                if (registrar.activity().isFinishing() || registrar.activity().isDestroyed())
                    return;
                if (ads.get(adUnitId) == null) {
                    ads.put(adUnitId, new MoPubInterstitial(registrar.activity(), adUnitId));
                }
                ad = ads.get(adUnitId);
                if (ad != null) {
                    final MethodChannel adChannel = new MethodChannel(registrar.messenger(),
                            MopubConstants.INTERSTITIAL_AD_CHANNEL + "_" + adUnitId);
                    ad.setInterstitialAdListener(createListener(adChannel));
                    if (!ad.isReady()) {
                        ad.load();
                    }
                }
                result.success(null);
                break;
            case MopubConstants.HAS_INTERSTITIAL_METHOD:
                ad = ads.get(adUnitId);
                if (ad != null) {
                    boolean isReady = ad.isReady();
                    result.success(isReady);
                    return;
                }
                result.success(false);
                break;
            case MopubConstants.DESTROY_INTERSTITIAL_METHOD:
                ad = ads.get(adUnitId);
                if (ad == null) {
                    return;
                } else {
                    ad.setInterstitialAdListener(null);
                    ad.destroy();
                    ads.remove(adUnitId);
                }
                result.success(null);
                break;
            default:
                result.notImplemented();
        }
    }

    private MoPubInterstitial.InterstitialAdListener createListener(final MethodChannel channel) {
        return new MoPubInterstitial.InterstitialAdListener() {
            @Override
            public void onInterstitialLoaded(MoPubInterstitial ad) {
                HashMap<String, Object> args = new HashMap<>();
                args.put("keywords", ad.getKeywords());

                channel.invokeMethod(MopubConstants.LOADED_METHOD, args);
            }

            @Override
            public void onInterstitialFailed(MoPubInterstitial ad, MoPubErrorCode errorCode) {
                HashMap<String, Object> args = new HashMap<>();
                args.put("keywords", ad.getKeywords());
                args.put("errorCode", errorCode.toString());

                channel.invokeMethod(MopubConstants.ERROR_METHOD, args);
            }

            @Override
            public void onInterstitialShown(MoPubInterstitial ad) {
                HashMap<String, Object> args = new HashMap<>();
                args.put("keywords", ad.getKeywords());

                channel.invokeMethod(MopubConstants.DISPLAYED_METHOD, args);
            }

            @Override
            public void onInterstitialClicked(MoPubInterstitial ad) {
                HashMap<String, Object> args = new HashMap<>();
                args.put("keywords", ad.getKeywords());

                channel.invokeMethod(MopubConstants.CLICKED_METHOD, args);
            }

            @Override
            public void onInterstitialDismissed(MoPubInterstitial ad) {
                HashMap<String, Object> args = new HashMap<>();
                args.put("keywords", ad.getKeywords());

                channel.invokeMethod(MopubConstants.DISMISSED_METHOD, args);
            }
        };
    }
}

File: mopub_flutter-1.2.2+1/ios/Classes/MopubInterstitialAdPlugin.swift
//
//  MopubInterstitialAdPlugin.swift
//  mopub_flutter
//
//  Created by Rohit Kulkarni on 1/6/20.
//

import Foundation
import Flutter
import MoPub

public class MopubInterstitialAdPlugin : NSObject, FlutterPlugin {
    
    fileprivate var ads: [String: MPInterstitialAdController] = [:]
    fileprivate var delegates: [String: MPInterstitialAdControllerDelegate] = [:]
    
    fileprivate var pluginRegistrar: FlutterPluginRegistrar?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = MopubInterstitialAdPlugin()
        instance.pluginRegistrar = registrar
        let channel = FlutterMethodChannel(name: MopubConstants.INTERSTITIAL_AD_CHANNEL, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    override init() {
        self.ads = Dictionary<String, MPInterstitialAdController>()
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String : Any] ?? [:]
        let adUnitId = args["adUnitId"] as? String ?? "ad_unit_id"
        
        switch call.method {
        case MopubConstants.SHOW_INTERSTITIAL_METHOD:
            let ad = ads[adUnitId] ?? nil
            if (ad == nil || !ad!.ready) {
                return;
            }
            let rootViewController = UIApplication.shared.keyWindow?.rootViewController
            ad!.show(from: rootViewController)
            result(nil);
            break
        case MopubConstants.LOAD_INTERSTITIAL_METHOD:
            
            if (ads[adUnitId] == nil) {
                ads[adUnitId] = MPInterstitialAdController(forAdUnitId: adUnitId);
            }
            let ad = ads[adUnitId]
            if (ad != nil) {
                let channelName = "\(MopubConstants.INTERSTITIAL_AD_CHANNEL)_\(adUnitId)"
                let adChannel = FlutterMethodChannel(name: channelName, binaryMessenger: pluginRegistrar!.messenger())
                delegates[adUnitId] = MopubInterstitialDelegate(channel: adChannel)
                ad!.delegate = delegates[adUnitId]
                if (!ad!.ready) {
                    ad!.loadAd()
                }
            }
            result(nil)
            break
        case MopubConstants.HAS_INTERSTITIAL_METHOD:
            let ad = ads[adUnitId]
            if(ad != nil) {
                let isReady = ad?.ready ?? false
                result(isReady)
                return
            }
            result(false)
            break
        case MopubConstants.DESTROY_INTERSTITIAL_METHOD:
            if (ads[adUnitId] == nil) {
                return;
            } else  {
                ads.removeValue(forKey: adUnitId)
                delegates.removeValue(forKey: adUnitId)
            }
            result(nil);
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

class MopubInterstitialDelegate: NSObject, MPInterstitialAdControllerDelegate {
    
    let channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    func interstitialDidLoadAd(_ interstitial: MPInterstitialAdController!) {
        var args:[String:Any] = [:]
        args["keywords"] = interstitial.keywords
        channel.invokeMethod(MopubConstants.LOADED_METHOD, arguments: args)
    }
    
    func interstitialDidFail(toLoadAd interstitial: MPInterstitialAdController!, withError error: Error!) {
        var args:[String:Any] = [:]
        args["keywords"] = interstitial.keywords
        args["errorCode"] = error.localizedDescription
        channel.invokeMethod(MopubConstants.ERROR_METHOD, arguments: args)
    }
    
    func interstitialDidReceiveTapEvent(_ interstitial: MPInterstitialAdController!) {
        var args:[String:Any] = [:]
        args["keywords"] = interstitial.keywords
        channel.invokeMethod(MopubConstants.CLICKED_METHOD, arguments: args)
    }
    
    func interstitialWillAppear(_ interstitial: MPInterstitialAdController!) {        
        var args:[String:Any] = [:]
        args["keywords"] = interstitial.keywords
        channel.invokeMethod(MopubConstants.CLICKED_METHOD, arguments: args)
    }
    
    func interstitialDidDisappear(_ interstitial: MPInterstitialAdController!) {
        var args:[String:Any] = [:]
        args["keywords"] = interstitial.keywords
        channel.invokeMethod(MopubConstants.DISMISSED_METHOD, arguments: args)
    }
    
    func interstitialDidExpire(_ interstitial: MPInterstitialAdController!) {
        interstitial.loadAd()
    }
    
}
